src/Microsoft.VisualStudio.Editor.Razor/RazorDirectiveCompletionSourceProvider.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    [System.Composition.Shared]
+    [Export(typeof(IAsyncCompletionSourceProvider))]
+    [Name("Razor directive completion provider.")]
+    [ContentType(RazorLanguage.ContentType)]
+    internal class RazorDirectiveCompletionSourceProvider : IAsyncCompletionSourceProvider
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+
+        [ImportingConstructor]
+        public RazorDirectiveCompletionSourceProvider(ForegroundDispatcher foregroundDispatcher)
+        {
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+        }
+
+        public IAsyncCompletionSource GetOrCreate(ITextView textView)
```

Comment:
This makes it sound like we're expected to cache this. Is there a reason why we should/should not?

Reply:
👍 should definitely cache.